cpp/src/join/nested_loop_join.cuh

```diff
@@ -28,167 +29,153 @@
-                                  null_equality compare_nulls,
-                                  rmm::cuda_stream_view stream)
-{
-  // The `right` table is always used for the inner loop. We want to use the smaller table
-  // for the inner loop. Thus, if `left` is smaller than `right`, swap `left/right`.
-  if ((JoinKind == join_kind::INNER_JOIN) && (right.num_rows() > left.num_rows())) {
-    return get_base_nested_loop_join_indices(right, left, true, JoinKind, compare_nulls, stream);
+  // If the output size will be zero, we can return immediately.
+  if (join_size == 0) {
+    return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                          std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
   }
-  // Trivial left join case - exit early
-  if ((JoinKind == join_kind::LEFT_JOIN) && (right.num_rows() == 0)) {
-    return get_trivial_left_join_indices(left, stream);
+
+  rmm::device_scalar<size_type> write_index(0, stream);
+
+  auto left_indices  = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+  auto right_indices = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+
+  const auto& join_output_l = left_indices->data();
+  const auto& join_output_r = right_indices->data();
+  if (has_nulls) {
+    conditional_join<block_size, DEFAULT_JOIN_CACHE_SIZE, true>
+      <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+        *left_table,
+        *right_table,
+        KernelJoinKind,
+        compare_nulls,
+        join_output_l,
+        join_output_r,
+        write_index.data(),
+        plan.dev_plan,
+        join_size);
+  } else {
+    conditional_join<block_size, DEFAULT_JOIN_CACHE_SIZE, false>
+      <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+        *left_table,
+        *right_table,
+        KernelJoinKind,
+        compare_nulls,
+        join_output_l,
+        join_output_r,
+        write_index.data(),
+        plan.dev_plan,
+        join_size);
   }
 
-  auto left_table  = table_device_view::create(left, stream);
-  auto right_table = table_device_view::create(right, stream);
+  CHECK_CUDA(stream.value());
 
-  size_type estimated_size = estimate_nested_loop_join_output_size(
-    *left_table, *right_table, JoinKind, compare_nulls, stream);
+  auto join_indices = std::make_pair(std::move(left_indices), std::move(right_indices));
 
-  // If the estimated output size is zero, return immediately
-  if (estimated_size == 0) {
-    return std::make_pair(rmm::device_uvector<size_type>{0, stream},
-                          rmm::device_uvector<size_type>{0, stream});
+  // For full joins, get the indices in the right table that were not joined to
+  // by any row in the left table.
+  if (JoinKind == join_kind::FULL_JOIN) {
+    auto complement_indices = detail::get_left_join_indices_complement(
```

Comment:
```suggestion
    auto const complement_indices = detail::get_left_join_indices_complement(
```

Reply:
Unfortunately this variable can't be `const` because `concatenate_vector_pairs` does not accept `const` inputs, for the reason that either of the two inputs (which are unique pointers) may need to be moved to construct the return value.